maintainers_annotation/maintainers_annotation.go

```diff
@@ -0,0 +1,190 @@
+			"validateAndParseScorecardYmlFile requires exactly one argument to save maintainers annotation data")
+	}
+	ma, ok := args[0].(*MaintainersAnnotation)
+	if !ok {
+		return false, fmt.Errorf(
+			"validateAndParseScorecardYmlFile requires argument of type *MaintainersAnnotation")
+	}
+
+	// Exclude scorecard GitHub Action file
+	if strings.Contains(path, ".github") {
+		return true, nil
+	}
+
+	// Keep looking for files if content is empty
+	if len(content) == 0 {
+		return true, nil
+	}
+
+	// Found scorecard.yml file, parse file
+	err := parseScorecardYmlFile(ma, content)
+	if err != nil {
+		return false, err
+	}
+
+	return false, nil
+}
+
+// readScorecardYmlFromRepo reads the scorecard.yml file from the repo and returns its configurations.
+func readScorecardYmlFromRepo(repoClient clients.RepoClient) (MaintainersAnnotation, error) {
+	ma := MaintainersAnnotation{}
+	err := fileparser.OnMatchingFileContentDo(repoClient, fileparser.PathMatcher{
+		Pattern:       "scorecard.yml",
+		CaseSensitive: false,
+	}, validateAndParseScorecardYmlFile, &ma)
+
+	if err != nil {
+		return ma, fmt.Errorf("%w", err)
+	}
+
+	if ma.Exemptions == nil {
+		return ma, errNoScorecardYmlFile
+	}
+
+	return ma, nil
+}
+
+// GetMaintainersAnnotation reads the maintainers annotation from the repo, stored in scorecard.yml, and returns it.
+func GetMaintainersAnnotation(repoClient clients.RepoClient) (MaintainersAnnotation, error) {
+	// Read the scorecard.yml file and parse it into maintainers annotation
+	ma, err := readScorecardYmlFromRepo(repoClient)
+
+	// If scorecard.yml doesn't exist or file doesn't contain exemptions, ignore
+	if err == errNoScorecardYmlFile {
+		return MaintainersAnnotation{}, nil
+	}
+	// If an error happened while finding or parsing scorecard.yml file, then raise error
+	if err != nil {
+		return MaintainersAnnotation{}, fmt.Errorf("%w", err)
+	}
+
+	// Return maintainers annotation
+	return ma, nil
+}
+
+// IsCheckExempted verifies if a given check in the results is exempted in maintainers annotation.
+func (ma *MaintainersAnnotation) IsCheckExempted(check checker.CheckResult) (bool, []string) {
+	// If check has a maximum score, then there it doesn't make sense anymore to reason the check
+	// This may happen if the check score was once low but then the problem was fixed on Scorecard side
+	// or on the maintainers side
+	if check.Score == checker.MaxResultScore {
+		return false, nil
+	}
+
+	for _, exemption := range ma.Exemptions {
+		for _, checkName := range exemption.Checks {
+			if strings.EqualFold(checkName, strings.ToLower(check.Name)) {
+				return true, getAnnotations(exemption.Annotations)
+			}
+		}
+	}
+	return false, nil
+}
+
+// getAnnotations parses a group of annotations into annotation reasons.
+func getAnnotations(a []Annotation) []string {
+	var annotationReasons []string
+	for _, annotation := range a {
+		annotationReasons = append(annotationReasons, AnnotationReasonDoc[annotation.AnnotationReason])
```

Comment:
what happens when someone passes an annotation that isn't one of the ones in the map?

Reply:
Good question. I was planning on solving this while doing the tests. My suggestion is that it should be logged that the reason or check was not identified but not throw an error in the execution, just ignore the weird check / annotation.

Reply:
Not sure if this makes sense. Another option is to log an error saying it could not parse the annotations and return an empty maintainers annotation to not break the execution.

Reply:
generally, logging that we're ignoring/skipping an unknown annotation but reading the rest of the file seems fine